exp/controllers/rosaroleconfig_controller.go

```diff
@@ -0,0 +1,663 @@
+	if oidcID == "" {
+		return nil
+	}
+	runtime := rosacli.NewRuntime()
+	runtime.OCMClient = ocmClient
+	runtime.Reporter = (&rosa.Reporter{})
+
+	runtime.Logger = rosalogging.NewLogger()
+	runtime.AWSClient, err = aws.NewClient().Logger(runtime.Logger).Build()
+	if err != nil {
+		return fmt.Errorf("failed to create aws client: %w", err)
+	}
+
+	oidcConfig, err := runtime.OCMClient.GetOidcConfig(oidcID)
+	if err != nil {
+		return err
+	}
+
+	providers, err := runtime.AWSClient.ListOidcProviders("", oidcConfig)
+	if err != nil {
+		return err
+	}
+	for _, provider := range providers {
+		if strings.Contains(provider.Arn, oidcID) {
+			scope.RosaRoleConfig.Status.OIDCProviderARN = provider.Arn
+			return nil
+		}
+	}
+
+	runtime.Creator, err = runtime.AWSClient.GetCreator()
+	if err != nil {
+		return err
+	}
+
+	return oidcprovider.CreateOIDCProvider(runtime, oidcID, "", true)
+}
+
+func (r *ROSARoleConfigReconciler) createAccountRoles(ctx context.Context, roleConfig *expinfrav1.ROSARoleConfig, scope *scope.RosaRoleConfigScope, ocmClient rosa.OCMClient) error {
+	config := roleConfig.Spec.AccountRoleConfig
+	runtime := rosacli.NewRuntime()
+	policies, err := ocmClient.GetPolicies("AccountRole")
+	if err != nil {
+		return err
+	}
+	runtime.OCMClient, err = rosa.NewOCMClient(ctx, scope)
+	if err != nil {
+		return err
+	}
+
+	runtime.Reporter = (&rosa.Reporter{})
+	runtime.Logger = rosalogging.NewLogger()
+	runtime.AWSClient, err = aws.NewClient().Logger(runtime.Logger).Build()
+	if err != nil {
+		return fmt.Errorf("failed to create aws client: %w", err)
+	}
+
+	runtime.Creator, err = runtime.AWSClient.GetCreator()
+	if err != nil {
+		return err
+	}
+
+	createRoles := true
+	accountRoles, err := runtime.AWSClient.ListAccountRoles(config.Version)
+	if err != nil {
+		// Let create account roles continue if no account roles are found
+		if !strings.Contains(err.Error(), "no account roles found") {
+			return err
+		}
+	}
+
+	for _, role := range accountRoles {
+		if role.RoleName == fmt.Sprintf("%s-HCP-ROSA-Installer-Role", config.Prefix) {
+			createRoles = false
+			scope.RosaRoleConfig.Status.AccountRolesRef.InstallerRoleARN = role.RoleARN
+		}
+		if role.RoleName == fmt.Sprintf("%s-HCP-ROSA-Support-Role", config.Prefix) {
+			createRoles = false
+			scope.RosaRoleConfig.Status.AccountRolesRef.SupportRoleARN = role.RoleARN
+		}
+		if role.RoleName == fmt.Sprintf("%s-HCP-ROSA-Worker-Role", config.Prefix) {
+			createRoles = false
+			scope.RosaRoleConfig.Status.AccountRolesRef.WorkerRoleARN = role.RoleARN
+		}
+	}
+	if createRoles {
+		runtime.Reporter = (&rosa.Reporter{})
+		runtime.Logger = rosalogging.NewLogger()
+		runtime.AWSClient, err = aws.NewClient().Logger(runtime.Logger).Build()
+		if err != nil {
+			return fmt.Errorf("failed to create aws client: %w", err)
+		}
+
+		managedPolicies := true
+		isSharedVpc := config.SharedVPCConfig.VPCEndpointRoleARN != "" && config.SharedVPCConfig.RouteRoleARN != ""
```

Comment:
This is used a few places in the code, might be worth making the check a helper function on config.SharedVPCConfig.